rspace/src/test/scala/coop/rchain/rspace/nextgenrspace/history/HistorySpec.scala

```diff
@@ -0,0 +1,138 @@
+package coop.rchain.rspace.nextgenrspace.history
+import coop.rchain.rspace.Blake2b256Hash
+import coop.rchain.rspace.test.TestKey32
+import org.scalatest.{FlatSpec, Matchers, OptionValues}
+
+import scala.collection.concurrent.TrieMap
+import History._
+import org.scalacheck.{Arbitrary, Gen, Shrink}
+import coop.rchain.rspace.test.ArbitraryInstances._
+import monix.eval.Task
+import org.scalatest.prop._
+import scala.concurrent.duration._
+import monix.execution.Scheduler.Implicits.global
+
+class HistorySpec
+    extends FlatSpec
+    with Matchers
+    with OptionValues
+    with GeneratorDrivenPropertyChecks
+    with InMemoryHistoryTestBase {
+
+  implicit def noShrink[T]: Shrink[T] = Shrink.shrinkAny
+
+  // disable shrinking since this test operates on a persistent structure
+  implicit val propertyCheckConfigParam = PropertyCheckConfiguration(minSuccessful = 1000)
+
+  type Key  = TestKey32
+  type Data = (Key, Blake2b256Hash)
+
+  "history" should "accept new leafs (insert, update, delete)" in forAll {
+    actions: List[(TestKey32, Blake2b256Hash)] =>
+      val emptyHistory =
+        new History[Task, TestKey32](emptyRootHash, inMemHistoryStore, inMemPointerBlockStore)
+
+      val emptyState = Map.empty[Key, (Data, History[Task, Key])] // accumulate actions performed on the trie
+
+      val (result, map) = actions.foldLeft((emptyHistory, emptyState)) {
+        case ((history, prevActions), action) =>
+          val (postHistory, _) =
+            insertAndVerify(history, action :: Nil, prevActions.values.map(_._1).toList)
+          val actions = prevActions + (action._1 -> (action, postHistory))
+          (postHistory, actions)
+      }
+
+      map.values.foreach {
+        case ((k, v), postModifyHistory) =>
+          val actions           = DeleteAction(k) :: Nil
+          val postDeleteHistory = postModifyHistory.process(actions).runSyncUnsafe(20.seconds)
+
+          fetchData(postModifyHistory, k)._1 shouldBe Leaf(v)
+          fetchData(postDeleteHistory, k)._1 shouldBe EmptyTrie
+      }
+
+      val deletions    = map.values.map(_._1).map(v => DeleteAction(v._1)).toList
+      val finalHistory = result.process(deletions)
+      finalHistory.runSyncUnsafe(20.seconds).root shouldBe emptyHistory.root
+  }
+
+  val arbitraryTestKey32: Arbitrary[TestKey32] =
+    Arbitrary(Gen.sized { _ =>
+      Gen
+        .listOfN(32, Arbitrary.arbitrary[Int])
+        .map(ints => TestKey32.create(ints))
+    })
+
+  val arbitrary2TestKey32: Arbitrary[TestKey32] =
+    Arbitrary(Gen.sized { _ =>
+      Gen
+        .listOfN(3, Arbitrary.arbitrary[Int])
+        .map(ints => TestKey32.create(List.fill(29)(0) ++ ints))
```

Comment:
what's the gain?

Reply:
Us not assuming a specific generator in tests, and the actual ability to test the 'top parts' of the trie and its edge cases (should there be any)

Reply:
the ability is still there, the test just uses a particular version of the key.

I can add a second test which uses a broader key.

Reply:
That was not my intention, but I won't be stopping you :)

Reply:
In my opinion "Us not assuming a specific generator in tests" is pointless. there is no gain in having abstract tests in a case where the possible extensions consist of using a different key (which no one should ever do, we want a working system for keys that align with production, not for shorter ones).

Reply:
I wouldn't be proposing polymorphism if I didn't see potential gains in testability and understandability and non-attachment-to-the-impls of that code. The latter two area at a tradeoff with having it concrete -> easier to grok the whole.

The gain in testability is: it's easier and more robust to grab a default Arbitrary for a smaller type than to tinker with the generator to test at the bottom of the tree.

I don't care, really. You asked, so I explained myself. IMO "pointless" is a bit harsh, and not really (I hope I explained why).

Reply:
I feel we're running into deep waters of philosophical debate.
let's agree that it's more important to have a close to production test now and I will drop a ticket to revisit this topic.

Reply:
https://rchain.atlassian.net/browse/RCHAIN-3229

Reply:
That ticket is only needed if you feel my point makes a difference :) I'm aware that it might not after the code has been written with a different approach :)